forge/src/main/java/org/openjdk/skara/forge/gitlab/GitLabRepository.java

```diff
@@ -170,10 +175,43 @@ public List<PullRequest> openPullRequestsAfter(ZonedDateTime updatedAfter) {
                 .param("updated_after", updatedAfter.format(DateTimeFormatter.ISO_DATE_TIME))
                 .execute().stream()
                 .filter(this::hasHeadHash)
+                .map(this::refetchMergeRequest)
                 .map(value -> new GitLabMergeRequest(this, gitLabHost, value, request))
                 .collect(Collectors.toList());
     }
 
+    /**
+     * This method is used to work around a bug in GitLab where list query
+     * results for merge requests sometimes return stale data. Fetching them
+     * directly using the ID will always return up-to-date data. The method
+     * logs when stale data is actually detected to give us a way to
+     * empirically verify when the bug is no longer present.
+     */
+    private JSONValue refetchMergeRequest(JSONValue origData) {
+        var updatedAt = ZonedDateTime.parse(origData.get("updated_at").asString());
+        // Only do the refetch on merge requests that have been updated recently.
+        // The 3 hours cut off is rather arbitrarily chosen. We will have to see
+        // if it is enough. Having some kind of cut off is reasonable as we would
+        // otherwise risk running a lot of queries on the first run after a
+        // restart.
+        if (updatedAt.isAfter(ZonedDateTime.now().minus(Duration.ofHours(3)))) {
+            var id = origData.get("iid");
+            var newData = request.get("merge_requests/" + id).execute();
+            // We can't compare the full json object returned from a list query
+            // and get query call as they will always be different. The part we
```

Comment:
They don't differ in any way that we care about when processing merge requests. We aren't looking at any of those fields. The only thing I can imagine this affecting is the PullRequestPoller, which may, at least in theory, detect some false positives due to this when detecting if a merge request has been updated or not. So that could potentially happen once 3 hours after a merge request has had its updated_at field updated, but only if the query was still including that 3 hour old result, which it normally shouldn't.